File: 1.1_before_major_refactor_2.0/csharp/core/Project.cs
using System;
using System.Collections.Generic;
using System.Xml;
using urakawa.progress;
using urakawa.property.channel;
using urakawa.metadata;
using urakawa.xuk;
using urakawa.events;
using urakawa.events.project;

namespace urakawa
{
	/// <summary>
	/// Represents a projects - part of the facade API, provides methods for opening and saving XUK files
	/// </summary>
	public class Project : XukAble, IValueEquatable<Project>, IChangeNotifier
	{
		
		#region Event related members

		/// <summary>
		/// Event fired after the <see cref="Project"/> has changed. 
		/// The event fire before any change specific event 
		/// </summary>
		public event EventHandler<DataModelChangedEventArgs> changed;
		/// <summary>
		/// Fires the <see cref="changed"/> event 
		/// </summary>
		/// <param name="args">The arguments of the event</param>
		protected void notifyChanged(DataModelChangedEventArgs args)
		{
			EventHandler<urakawa.events.DataModelChangedEventArgs> d = changed;
			if (d != null) d(this, args);
		}
		/// <summary>
		/// Event fired after a <see cref="Presentation"/> has been added to the <see cref="Project"/>
		/// </summary>
		public event EventHandler<PresentationAddedEventArgs> presentationAdded;
		/// <summary>
		/// Fires the <see cref="presentationAdded"/> event
		/// </summary>
		/// <param name="source">
		/// The source, that is the <see cref="Project"/> to which a <see cref="Presentation"/> was added
		/// </param>
		/// <param name="addedPres">The <see cref="Presentation"/> that was added</param>
		protected void notifyPresentationAdded(Project source, Presentation addedPres)
		{
			EventHandler<PresentationAddedEventArgs> d = presentationAdded;
			if (d != null) d(this, new PresentationAddedEventArgs(source, addedPres));
		}

		void this_presentationAdded(object sender, PresentationAddedEventArgs e)
		{
			notifyChanged(e);
			e.AddedPresentation.changed += new EventHandler<DataModelChangedEventArgs>(Presentation_changed);
		}

		void Presentation_changed(object sender, DataModelChangedEventArgs e)
		{
			notifyChanged(e);
		}

		/// <summary>
		/// Event fired after a <see cref="Presentation"/> has been added to the <see cref="Project"/>
		/// </summary>
		public event EventHandler<PresentationRemovedEventArgs> presentationRemoved;
		/// <summary>
		/// Fires the <see cref="presentationRemoved"/> event
		/// </summary>
		/// <param name="source">
		/// The source, that is the <see cref="Project"/> to which a <see cref="Presentation"/> was added
		/// </param>
		/// <param name="removedPres">The <see cref="Presentation"/> that was added</param>
		protected void notifyPresentationRemoved(Project source, Presentation removedPres)
		{
			EventHandler<PresentationRemovedEventArgs> d = presentationRemoved;
			if (d != null) d(this, new PresentationRemovedEventArgs(source, removedPres));
		}

		void this_presentationRemoved(object sender, PresentationRemovedEventArgs e)
		{
			e.RemovedPresentation.changed -= new EventHandler<DataModelChangedEventArgs>(Presentation_changed);
			notifyChanged(e);
		}
		#endregion


		private List<Presentation> mPresentations;


		/// <summary>
		/// Default constructor
		/// </summary>
		public Project()
		{
			mPresentations = new List<Presentation>();
			this.presentationAdded += new EventHandler<PresentationAddedEventArgs>(this_presentationAdded);
			this.presentationRemoved += new EventHandler<PresentationRemovedEventArgs>(this_presentationRemoved);
		}

		private DataModelFactory mDataModelFactory;

		/// <summary>
		/// Gets the <see cref="DataModelFactory"/> of the <see cref="Project"/>
		/// </summary>
		/// <returns>The factory</returns>
		/// <remarks>
		/// The <see cref="DataModelFactory"/> of a <see cref="Project"/> is initialized lazily,
		/// in that if the <see cref="DataModelFactory"/> has not been explicitly initialized
		/// using the <see cref="setDataModelFactory"/>, then calling <see cref="getDataModelFactory"/>
		/// will initialize the <see cref="Project"/> with a newly created <see cref="DataModelFactory"/>.
		/// </remarks>
		public DataModelFactory getDataModelFactory()
		{
			if (mDataModelFactory == null) mDataModelFactory = new DataModelFactory();
			return mDataModelFactory;
		}

		/// <summary>
		/// Initializes the <see cref="Project"/> with a <see cref="DataModelFactory"/>
		/// </summary>
		/// <param name="fact">The factory with which to initialize - must not be <c>null</c></param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref name="fact"/> is <c>null</c>
		/// </exception>
		/// <exception cref="exception.IsAlreadyInitializedException">
		/// Thrown when the <see cref="Project"/> has already been initialized with a <see cref="DataModelFactory"/>
		/// </exception>
		public void setDataModelFactory(DataModelFactory fact)
		{
			if (fact == null)
			{
				throw new exception.MethodParameterIsNullException(
					"The DataModelFactory can not be null");
			}
			if (mDataModelFactory != null)
			{
				throw new exception.IsAlreadyInitializedException(
					"The Project has already been initialized with a DataModelFactory");
			}
			mDataModelFactory = fact;
		}



		/// <summary>
		/// Opens an XUK file and loads the project from this
		/// </summary>
		/// <param name="fileUri">The <see cref="Uri"/> of the source XUK file</param>
		public void openXUK(Uri fileUri)
		{
            OpenXukAction action = new OpenXukAction(fileUri, this);
            action.execute();
		}

		/// <summary>
		/// Saves the <see cref="Project"/> to a XUK file
		/// </summary>
		/// <param name="fileUri">The <see cref="Uri"/> of the destination XUK file</param>
		public void saveXUK(Uri fileUri)
		{
            SaveXukAction action = new SaveXukAction(fileUri, this);
            action.execute();
		}



		/// <summary>
		/// Gets the <see cref="urakawa.Presentation"/> of the <see cref="Project"/> at a given index
		/// </summary>
		/// <param name="index">The index of the <see cref="Presentation"/> to get</param>
		/// <returns>The presentation at the given index</returns>
		/// <exception cref="exception.MethodParameterIsOutOfBoundsException">
		/// Thrown when <paramref name="index"/> is not in <c>[0;this.getNumberOfPresentations()-1]</c>
		/// </exception>
		public Presentation getPresentation(int index)
		{
			if (index<0 || getNumberOfPresentations()<=index)
			{
				throw new exception.MethodParameterIsOutOfBoundsException(String.Format(
					"There is no Presentation at index {0:0}, index must be between 0 and {1:0}",
					index, getNumberOfPresentations()-1));
			}
			return mPresentations[index];
		}

		/// <summary>
		/// Gets a list of the <see cref="Presentation"/>s in the <see cref="Project"/>
		/// </summary>
		/// <returns>The list</returns>
		public List<Presentation> getListOfPresentations()
		{
			return new List<Presentation>(mPresentations);
		}

		/// <summary>
		/// Sets the <see cref="Presentation"/> at a given index
		/// </summary>
		/// <param name="newPres">The <see cref="Presentation"/> to set</param>
		/// <param name="index">The given index</param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref name="newPres"/> is <c>null</c></exception>
		/// <exception cref="exception.MethodParameterIsOutOfBoundsException">
		/// Thrown when <paramref name="index"/> is not in <c>[0;this.getNumberOfPresentations()]</c>
		/// </exception>
		/// <exception cref="exception.IsAlreadyManagerOfException">
		/// Thrown when <paramref name="newPres"/> already exists in <c>this</c> with another <paramref name="index"/>
		/// </exception>
		public void setPresentation(Presentation newPres, int index)
		{
			if (newPres == null)
			{
				throw new exception.MethodParameterIsNullException("The new Presentation can not be null");
			}
			if (index < 0 || getNumberOfPresentations() < index)
			{
				throw new exception.MethodParameterIsOutOfBoundsException(String.Format(
					"There is no Presentation at index {0:0}, index must be between 0 and {1:0}",
					index, getNumberOfPresentations()));
			}
			if (mPresentations.Contains(newPres))
			{
				if (mPresentations.IndexOf(newPres) != index)
				{
					throw new exception.IsAlreadyManagerOfException(String.Format(
						"The new Presentation already exists in the Project at index {0:0}",
						mPresentations.IndexOf(newPres)));
				}
			}
			if (index < getNumberOfPresentations())
			{
				removePresentation(index);
				mPresentations.Insert(index, newPres);
			}
			else
			{
				mPresentations.Add(newPres);
			}
			newPres.setProject(this);
			notifyPresentationAdded(this, newPres);
		}

		/// <summary>
		/// Adds a <see cref="Presentation"/> to the <see cref="Project"/>
		/// </summary>
		/// <param name="newPres">The <see cref="Presentation"/> to add</param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref name="newPres"/> is <c>null</c></exception>
		/// <exception cref="exception.IsAlreadyManagerOfException">
		/// Thrown when <paramref name="newPres"/> already exists in <c>this</c>
		/// </exception>
		public void addPresentation(Presentation newPres)
		{
			setPresentation(newPres, getNumberOfPresentations());
		}

		/// <summary>
		/// Adds a newly created <see cref="Presentation"/> to the <see cref="Project"/>,
		/// as returned by <c><see cref="getDataModelFactory"/>().<see cref="DataModelFactory.createPresentation()"/>()</c>
		/// </summary>
		/// <returns>The newly created and added <see cref="Presentation"/></returns>
		public Presentation addNewPresentation()
		{
			Presentation newPres = getDataModelFactory().createPresentation();
			addPresentation(newPres);
			return newPres;
		}

		/// <summary>
		/// Gets the number of <see cref="Presentation"/>s in the <see cref="Project"/>
		/// </summary>
		/// <returns>The number of <see cref="Presentation"/>s</returns>
		public int getNumberOfPresentations()
		{
			return mPresentations.Count;
		}

		/// <summary>
		/// Removes the <see cref="Presentation"/> at a given index
		/// </summary>
		/// <param name="index">The given index</param>
		/// <returns>The removed <see cref="Presentation"/></returns>
		/// <exception cref="exception.MethodParameterIsOutOfBoundsException">
		/// Thrown when <paramref name="index"/> is not in <c>[0;this.getNumberOfPresentations()-1]</c>
		/// </exception>
		public Presentation removePresentation(int index)
		{
			if (index < 0 || getNumberOfPresentations() <= index)
			{
				throw new exception.MethodParameterIsOutOfBoundsException(String.Format(
					"There is no Presentation at index {0:0}, index must be between 0 and {1:0}",
					index, getNumberOfPresentations() - 1));
			}
			Presentation pres = getPresentation(index);
			mPresentations.RemoveAt(index);
			notifyPresentationRemoved(this, pres);
			return pres;
		}

		/// <summary>
		/// Removes all <see cref="Presentation"/>s from the <see cref="Project"/>
		/// </summary>
		public void removeAllPresentations()
		{
			mPresentations.Clear();
		}

		
		#region IXUKAble members

		/// <summary>
		/// Clears the <see cref="Project"/>, removing all <see cref="Presentation"/>s
		/// </summary>
		protected override void clear()
		{
			removeAllPresentations();
			base.clear();
		}

		/// <summary>
		/// Reads a child of a Project xuk element. 
		/// </summary>
		/// <param name="source">The source <see cref="XmlReader"/></param>
        /// <param name="handler">The handler for progress</param>
        protected override void xukInChild(XmlReader source, ProgressHandler handler)
		{
			bool readItem = false;
			if (source.NamespaceURI == ToolkitSettings.XUK_NS)
			{
				switch (source.LocalName)
				{
					case "mPresentations":
						xukInPresentations(source, handler);
						readItem = true;
						break;
				}
			}
			if (!readItem) base.xukInChild(source, handler);
		}

		private void xukInPresentations(XmlReader source, ProgressHandler handler)
		{
			if (!source.IsEmptyElement)
			{
				while (source.Read())
				{
					if (source.NodeType == XmlNodeType.Element)
					{
						Presentation pres = getDataModelFactory().createPresentation(
							source.LocalName, source.NamespaceURI);
						if (pres != null)
						{
							this.addPresentation(pres);
							pres.xukIn(source, handler);
						}
						else if (!source.IsEmptyElement)
						{
							source.ReadSubtree().Close();
						}
					}
					else if (source.NodeType == XmlNodeType.EndElement)
					{
						break;
					}
					if (source.EOF) throw new exception.XukException("Unexpectedly reached EOF");
				}
			}
		}

		/// <summary>
		/// Write the child elements of a Project element.
		/// </summary>
		/// <param name="destination">The destination <see cref="XmlWriter"/></param>
		/// <param name="baseUri">
		/// The base <see cref="Uri"/> used to make written <see cref="Uri"/>s relative, 
		/// if <c>null</c> absolute <see cref="Uri"/>s are written
		/// </param>
        /// <param name="handler">The handler for progress</param>
        protected override void xukOutChildren(XmlWriter destination, Uri baseUri, ProgressHandler handler)
		{
			base.xukOutChildren(destination, baseUri, handler);
			destination.WriteStartElement("mPresentations", ToolkitSettings.XUK_NS);
			foreach (Presentation pres in getListOfPresentations())
			{
				pres.xukOut(destination, baseUri, handler);
			}
			destination.WriteEndElement();
		}

		#endregion



		#region IValueEquatable<Project> Members

		/// <summary>
		/// Determines of <c>this</c> has the same value as a given other instance
		/// </summary>
		/// <param name="other">The other instance</param>
		/// <returns>A <see cref="bool"/> indicating the result</returns>
		public bool valueEquals(Project other)
		{
			if (other==null) return false;
			if (GetType()!=other.GetType()) return false;
			if (getNumberOfPresentations()!=other.getNumberOfPresentations()) return false;
			for (int index=0; index<getNumberOfPresentations(); index++)
			{
				if (!getPresentation(index).valueEquals(other.getPresentation(index))) return false;
			}
			return true;
		}

		#endregion
	}
}

File: java/core/src/org/daisy/urakawa/Project.java
package org.daisy.urakawa;

import java.net.URI;
import java.util.LinkedList;
import java.util.List;

import org.daisy.urakawa.command.CommandCannotExecuteException;
import org.daisy.urakawa.events.DataModelChangedEvent;
import org.daisy.urakawa.events.Event;
import org.daisy.urakawa.events.EventHandler;
import org.daisy.urakawa.events.IEventHandler;
import org.daisy.urakawa.events.IEventListener;
import org.daisy.urakawa.events.project.PresentationAddedEvent;
import org.daisy.urakawa.events.project.PresentationRemovedEvent;
import org.daisy.urakawa.exception.IsAlreadyInitializedException;
import org.daisy.urakawa.exception.IsAlreadyManagerOfException;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.exception.MethodParameterIsOutOfBoundsException;
import org.daisy.urakawa.nativeapi.IXmlDataReader;
import org.daisy.urakawa.nativeapi.IXmlDataWriter;
import org.daisy.urakawa.progress.IProgressHandler;
import org.daisy.urakawa.progress.ProgressCancelledException;
import org.daisy.urakawa.xuk.AbstractXukAble;
import org.daisy.urakawa.xuk.IXukAble;
import org.daisy.urakawa.xuk.OpenXukAction;
import org.daisy.urakawa.xuk.SaveXukAction;
import org.daisy.urakawa.xuk.XukDeserializationFailedException;
import org.daisy.urakawa.xuk.XukSerializationFailedException;

/**
 * Reference implementation of the interface.
 * 
 * @composed - - 1 org.daisy.urakawa.PresentationFactory
 * @composed 1 - 0..n org.daisy.urakawa.Presentation
 */
public class Project extends AbstractXukAble implements IWithPresentations,
        IValueEquatable<Project>, IEventHandler<DataModelChangedEvent>
{
    private PresentationFactory mPresentationFactory;
    private List<Presentation> mPresentations;
    // The 2 event bus below handle events related to adding and removing
    // presentations to and from this project.
    // Please note that this class automatically adds a listener for each bus,
    // in order to handle
    // the (de)registration of a special listener (mBubbleEventListener) which
    // forwards the bubbling events from the Data Model. See comment for
    // mBubbleEventListener.
    protected IEventHandler<Event> mPresentationAddedEventNotifier = new EventHandler();
    protected IEventHandler<Event> mPresentationRemovedEventNotifier = new EventHandler();
    // This event bus receives all the events that are raised from within the
    // Data Model of the underlying Presentations of this Project, including
    // the
    // above built-in events (PresentationRemoved and
    // PresentationRemoved).
    protected IEventHandler<Event> mDataModelEventNotifier = new EventHandler();

    // This "hub" method automatically dispatches the notify() call to the
    // appropriate IEventHandler (either mPresentationAddedEventNotifier,
    // mPresentationRemovedEventNotifier or mDataModelEventNotifier), based on
    // the type of the given event. Please note that the PresentationAdded and
    // PresentationRemoved events are passed to the generic
    // mDataModelEventNotifier event bus as well as to their corresponding
    // mPresentationAddedEventNotifier and mPresentationRemovedEventNotifier
    // bus.
    /**
     * @hidden
     */
    public <K extends DataModelChangedEvent> void notifyListeners(K event)
            throws MethodParameterIsNullException
    {
        if (event == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (PresentationAddedEvent.class.isAssignableFrom(event.getClass()))
        {
            mPresentationAddedEventNotifier.notifyListeners(event);
        }
        else
            if (PresentationRemovedEvent.class.isAssignableFrom(event
                    .getClass()))
            {
                mPresentationRemovedEventNotifier.notifyListeners(event);
            }
        mDataModelEventNotifier.notifyListeners(event);
    }

    // This "hub" method automatically dispatches the registerListener() call to
    // the
    // appropriate IEventHandler (either mPresentationAddedEventNotifier,
    // mPresentationRemovedEventNotifier or mDataModelEventNotifier), based on
    // the class type given. Please note that the PresentationAdded and
    // PresentationRemoved listeners are not registered with the generic
    // mDataModelEventNotifier event bus (only to their corresponding
    // mPresentationAddedEventNotifier and mPresentationRemovedEventNotifier
    // bus).
    /**
     * @hidden
     */
    public <K extends DataModelChangedEvent> void registerListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (listener == null || klass == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (PresentationAddedEvent.class.isAssignableFrom(klass))
        {
            mPresentationAddedEventNotifier.registerListener(listener, klass);
        }
        else
            if (PresentationRemovedEvent.class.isAssignableFrom(klass))
            {
                mPresentationRemovedEventNotifier.registerListener(listener,
                        klass);
            }
            else
            {
                mDataModelEventNotifier.registerListener(listener, klass);
            }
    }

    // Same as above, for de-registration.
    /**
     * @hidden
     */
    public <K extends DataModelChangedEvent> void unregisterListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (listener == null || klass == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (PresentationAddedEvent.class.isAssignableFrom(klass))
        {
            mPresentationAddedEventNotifier.unregisterListener(listener, klass);
        }
        else
            if (PresentationRemovedEvent.class.isAssignableFrom(klass))
            {
                mPresentationRemovedEventNotifier.unregisterListener(listener,
                        klass);
            }
            else
            {
                mDataModelEventNotifier.unregisterListener(listener, klass);
            }
    }

    // This listener receives events that are raised from within the
    // Presentations of this Project.
    // It simply forwards the received event to the main event bus for this
    // Project (which by default has no registered listeners: application
    // programmers should manually register their listeners by calling
    // Project.registerListener(IEventListener<DataModelChangedEvent>,
    // DataModelChangedEvent.class)), or
    // Project.registerListener(IEventListener<ChildAddedEvent>,
    // ChildAddedEvent.class)), or
    // Project.registerListener(IEventListener<PresentationRemovedEvent>,
    // PresentationRemovedEvent.class)), etc.
    protected IEventListener<DataModelChangedEvent> mBubbleEventListener = new IEventListener<DataModelChangedEvent>()
    {
        public <K extends DataModelChangedEvent> void eventCallback(K event)
                throws MethodParameterIsNullException
        {
            if (event == null)
            {
                throw new MethodParameterIsNullException();
            }
            notifyListeners(event);
        }
    };
    // This built-in listener takes care of registering the
    // mBubbleEventListener for a Presentation when that Presentation is added
    // to the Project.
    protected IEventListener<PresentationAddedEvent> mPresentationAddedEventListener = new IEventListener<PresentationAddedEvent>()
    {
        public <K extends PresentationAddedEvent> void eventCallback(K event)
                throws MethodParameterIsNullException
        {
            if (event == null)
            {
                throw new MethodParameterIsNullException();
            }
            if (event.getSourceProject() == Project.this)
            {
                event.getAddedPresentation().registerListener(
                        mBubbleEventListener, DataModelChangedEvent.class);
            }
            else
            {
                throw new RuntimeException("WFT ??! This should never happen.");
            }
        }
    };
    // This built-in listener takes care of unregistering the
    // mBubbleEventListener for a Presentation when that Presentation is
    // removed
    // from the Project.
    protected IEventListener<PresentationRemovedEvent> mPresentationRemovedEventListener = new IEventListener<PresentationRemovedEvent>()
    {
        public <K extends PresentationRemovedEvent> void eventCallback(K event)
                throws MethodParameterIsNullException
        {
            if (event == null)
            {
                throw new MethodParameterIsNullException();
            }
            if (event.getSourceProject() == Project.this)
            {
                event.getRemovedPresentation().unregisterListener(
                        mBubbleEventListener, DataModelChangedEvent.class);
            }
            else
            {
                throw new RuntimeException("WFT ??! This should never happen.");
            }
        }
    };

    /**
     * Default constructor (empty list of Presentations)
     */
    public Project()
    {
        mPresentations = new LinkedList<Presentation>();
        try
        {
            registerListener(mPresentationAddedEventListener,
                    PresentationAddedEvent.class);
            registerListener(mPresentationRemovedEventListener,
                    PresentationRemovedEvent.class);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }

    /**
     * @return
     * @hidden
     */
    public PresentationFactory getPresentationFactory()
    {
        if (mPresentationFactory == null)
        {
            try
            {
                setPresentationFactory(new PresentationFactory());
            }
            catch (MethodParameterIsNullException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e);
            }
            catch (IsAlreadyInitializedException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e);
            }
        }
        return mPresentationFactory;
    }

    /**
     * @hidden
     */
    private void setPresentationFactory(PresentationFactory factory)
            throws MethodParameterIsNullException,
            IsAlreadyInitializedException
    {
        if (factory == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (mPresentationFactory != null)
        {
            throw new IsAlreadyInitializedException();
        }
        mPresentationFactory = factory;
    }

    /**
     * <p>
     * Reads a XUK-formatted XML file, and generates the equivalent object data
     * that makes the Project.
     * </p>
     * 
     * @param uri
     *        cannot be null.
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    public void openXUK(URI uri) throws MethodParameterIsNullException
    {
        if (uri == null)
        {
            throw new MethodParameterIsNullException();
        }
        OpenXukAction action = new OpenXukAction(uri, this);
        try
        {
            action.execute();
        }
        catch (CommandCannotExecuteException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
    }

    /**
     * <p>
     * Writes the object data of the Project into a XUK-formatted XML file.
     * </p>
     * 
     * @param uri
     *        cannot be null
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    public void saveXUK(URI uri) throws MethodParameterIsNullException
    {
        if (uri == null)
        {
            throw new MethodParameterIsNullException();
        }
        SaveXukAction action = new SaveXukAction(uri, this);
        try
        {
            action.execute();
        }
        catch (CommandCannotExecuteException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
    }

    /**
     * @hidden
     */
    public boolean ValueEquals(Project other)
            throws MethodParameterIsNullException
    {
        if (other == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (getClass() != other.getClass())
            return false;
        if (getNumberOfPresentations() != other.getNumberOfPresentations())
            return false;
        for (int index = 0; index < getNumberOfPresentations(); index++)
        {
            try
            {
                if (!getPresentation(index).ValueEquals(
                        other.getPresentation(index)))
                    return false;
            }
            catch (MethodParameterIsOutOfBoundsException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ?!", e);
            }
        }
        return true;
    }

    /**
     * This method calls {@link org.daisy.urakawa.Presentation#cleanup()} for
     * each owned Presentation.
     */
    public void cleanup()
    {
        /**
         * Does nothing.
         */
    }

    /**
     * @hidden
     */
    public Presentation addNewPresentation()
    {
        Presentation newPres = getPresentationFactory().create();
        try
        {
            addPresentation(newPres);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
        catch (IsAlreadyManagerOfException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
        return newPres;
    }

    /**
     * @hidden
     */
    public void addPresentation(Presentation iPresentation)
            throws MethodParameterIsNullException, IsAlreadyManagerOfException
    {
        try
        {
            setPresentation(iPresentation, getNumberOfPresentations());
        }
        catch (MethodParameterIsOutOfBoundsException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
    }

    /**
     * @hidden
     */
    public List<Presentation> getListOfPresentations()
    {
        return new LinkedList<Presentation>(mPresentations);
    }

    /**
     * @hidden
     */
    public int getNumberOfPresentations()
    {
        return mPresentations.size();
    }

    /**
     * @hidden
     */
    public Presentation getPresentation(int index)
            throws MethodParameterIsOutOfBoundsException
    {
        if (index < 0 || getNumberOfPresentations() <= index)
        {
            throw new MethodParameterIsOutOfBoundsException();
        }
        return mPresentations.get(index);
    }

    /**
     * @hidden
     */
    public void removeAllPresentations()
    {
        mPresentations.clear();
    }

    /**
     * @hidden
     */
    public Presentation removePresentation(int index)
            throws MethodParameterIsOutOfBoundsException
    {
        if (index < 0 || getNumberOfPresentations() <= index)
        {
            throw new MethodParameterIsOutOfBoundsException();
        }
        Presentation pres = getPresentation(index);
        mPresentations.remove(index);
        try
        {
            notifyListeners(new PresentationRemovedEvent(this, pres));
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
        return pres;
    }

    /**
     * @hidden
     */
    public void setPresentation(Presentation iPresentation, int index)
            throws MethodParameterIsNullException,
            MethodParameterIsOutOfBoundsException, IsAlreadyManagerOfException
    {
        if (iPresentation == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (index < 0 || getNumberOfPresentations() < index)
        {
            throw new MethodParameterIsOutOfBoundsException();
        }
        if (mPresentations.contains(iPresentation))
        {
            if (mPresentations.indexOf(iPresentation) != index)
            {
                throw new IsAlreadyManagerOfException();
            }
        }
        if (index < getNumberOfPresentations())
        {
            removePresentation(index);
            mPresentations.add(index, iPresentation);
        }
        else
        {
            mPresentations.add(iPresentation);
        }
        try
        {
            iPresentation.setProject(this);
        }
        catch (IsAlreadyInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        notifyListeners(new PresentationAddedEvent(this, iPresentation));
    }

    /**
     * @hidden
     */
    private void xukInPresentations(IXmlDataReader source, IProgressHandler ph)
            throws XukDeserializationFailedException,
            MethodParameterIsNullException, ProgressCancelledException
    {
        if (source == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        if (!source.isEmptyElement())
        {
            while (source.read())
            {
                if (source.getNodeType() == IXmlDataReader.ELEMENT)
                {
                    Presentation pres = null;
                    try
                    {
                        pres = getPresentationFactory()
                                .create(source.getLocalName(),
                                        source.getNamespaceURI());
                    }
                    catch (MethodParameterIsNullException e)
                    {
                        // Should never happen
                        throw new RuntimeException("WTF ??!", e);
                    }
                    catch (MethodParameterIsEmptyStringException e)
                    {
                        // Should never happen
                        throw new RuntimeException("WTF ??!", e);
                    }
                    if (pres != null)
                    {
                        try
                        {
                            addPresentation(pres);
                        }
                        catch (MethodParameterIsNullException e)
                        {
                            // Should never happen
                            throw new RuntimeException("WTF ??!", e);
                        }
                        catch (IsAlreadyManagerOfException e)
                        {
                            // Should never happen
                            throw new RuntimeException("WTF ??!", e);
                        }
                        try
                        {
                            pres.xukIn(source, ph);
                        }
                        catch (MethodParameterIsNullException e)
                        {
                            // Should never happen
                            throw new RuntimeException("WTF ??!", e);
                        }
                    }
                    else
                    {
                        super.xukInChild(source, ph);
                    }
                }
                else
                    if (source.getNodeType() == IXmlDataReader.END_ELEMENT)
                    {
                        break;
                    }
                if (source.isEOF())
                {
                    throw new XukDeserializationFailedException();
                }
            }
        }
    }

    /**
     * @hidden
     */
    @Override
    protected void clear()
    {
        removeAllPresentations();
        // super.clear();
    }

    /**
     * @hidden
     */
    @SuppressWarnings("unused")
    @Override
    protected void xukInAttributes(IXmlDataReader source, IProgressHandler ph)
            throws MethodParameterIsNullException,
            XukDeserializationFailedException, ProgressCancelledException
    {
        if (source == null)
        {
            throw new MethodParameterIsNullException();
        }
        // To avoid event notification overhead, we bypass this:
        if (false && ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
    }

    /**
     * @hidden
     */
    @Override
    protected void xukInChild(IXmlDataReader source, IProgressHandler ph)
            throws MethodParameterIsNullException,
            XukDeserializationFailedException, ProgressCancelledException
    {
        if (source == null)
        {
            throw new MethodParameterIsNullException();
        }
        // To avoid event notification overhead, we bypass this:
        if (false && ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        boolean readItem = false;
        if (source.getNamespaceURI() == IXukAble.XUK_NS)
        {
            if (source.getLocalName() == "mPresentations")
            {
                try
                {
                    xukInPresentations(source, ph);
                }
                catch (MethodParameterIsNullException e)
                {
                    // Should never happen
                    throw new RuntimeException("WTF ??!", e);
                }
                readItem = true;
            }
        }
        if (!readItem)
        {
            super.xukInChild(source, ph);
        }
    }

    /**
     * @hidden
     */
    @SuppressWarnings("unused")
    @Override
    protected void xukOutAttributes(IXmlDataWriter destination, URI baseUri,
            IProgressHandler ph) throws XukSerializationFailedException,
            MethodParameterIsNullException, ProgressCancelledException
    {
        if (destination == null || baseUri == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
    }

    /**
     * @hidden
     */
    @Override
    protected void xukOutChildren(IXmlDataWriter destination, URI baseUri,
            IProgressHandler ph) throws XukSerializationFailedException,
            MethodParameterIsNullException, ProgressCancelledException
    {
        if (destination == null || baseUri == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        // super.xukOutChildren(destination, baseUri);
        destination.writeStartElement("mPresentations", IXukAble.XUK_NS);
        for (Presentation pres : getListOfPresentations())
        {
            pres.xukOut(destination, baseUri, ph);
        }
        destination.writeEndElement();
    }
}
